[PHIElimination] Ensure correct copy emission for `INLINEASM_BR` indirect targets

antoniofrighetto opened the pull request:
While lowering phi-nodes to copies, subsequently to f7a53d82c0902147909f28a9295a9d00b4b27d38, ensure that such copies are placed before any INLINEASM_BR instructions, only if this actually happens to be the case. If a block with an INLINEASM_BR instruction and any of its indirect jump target are distinct blocks, the new copy on the indirect target continues to be normally emitted prior to the terminator.

Fixes: https://github.com/llvm/llvm-project/issues/172824.

arsenm commented on llvm/test/CodeGen/X86/callbr-asm-different-indirect-target.mir:
```diff
@@ -0,0 +1,88 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 6
+# RUN: llc -mtriple=x86_64-- -verify-machineinstrs -o - %s -run-pass=livevars,phi-node-elimination | FileCheck %s
+# RUN: llc -mtriple=x86_64-- -verify-machineinstrs -o - %s --passes='require<live-vars>,phi-node-elimination' -verify-each | FileCheck %s
```
```suggestion
# RUN: llc -mtriple=x86_64-- -run-pass=livevars,phi-node-elimination %s -o - | FileCheck %s
# RUN: llc -mtriple=x86_64-- --passes='require<live-vars>,phi-node-elimination' -verify-each -o - %s | FileCheck %s
```

arsenm commented on llvm/test/CodeGen/X86/callbr-asm-different-indirect-target.mir:
```diff
@@ -0,0 +1,88 @@
+# RUN: llc -mtriple=x86_64-- -verify-machineinstrs -o - %s --passes='require<live-vars>,phi-node-elimination' -verify-each | FileCheck %s
+
+--- |
+  define i64 @inlineasm_br_different_indirect_target(i1 %cmp) {
```
Can you also add an end to end test

arsenm commented on llvm/lib/CodeGen/PHIEliminationUtils.cpp:
```diff
@@ -32,7 +32,21 @@ llvm::findPHICopyInsertPoint(MachineBasicBlock* MBB, MachineBasicBlock* SuccMBB,
+  // INLINEASM_BR, we may not need any specialized handling.
+  bool HasInlineAsmBrToSucc = false;
+  if (SuccMBB->isInlineAsmBrIndirectTarget()) {
+    for (MachineInstr &MI : *MBB) {
```
```suggestion
    for (const MachineInstr &MI : *MBB) {
```

nikic commented on llvm/lib/CodeGen/PHIEliminationUtils.cpp:
```diff
@@ -13,6 +13,19 @@
+  if (!SuccMBB->isInlineAsmBrIndirectTarget())
+    return false;
+
+  for (const MachineInstr &MI : *MBB)
```
Does this need to scan the whole block?

antoniofrighetto replied:
Not sure if there's a best alternative, INLINEASM_BR doesn't seem to be a terminator and we might introduce copies between INLINEASM_BR and the jump to its fallthrough during codegen.

qcolombet replied:
Should we reverse-walk the BB at least?
I'd expect the inlineasm br to be towards the end of the basic block.

antoniofrighetto replied:
> Should we reverse-walk the BB at least?

Moved to reverse walk (I wouldn't expect the indirect target BB to be the same one where the INLINEASM_BR lies on average, though better than nothing).

> It is required to be in the terminators

Tested whether INLINEASM_BR appears in `terminators()`, and it doesn't seem to be the case. I agree it should be (like its IR counterpart), I'll take a look as a follow-up.

nickdesaulniers replied:
> Does this need to scan the whole block?

Yes since INLINEASM_BR is not a Terminator.

XChy reviewed (approved):
LGTM. I may be able to investigate #156249 further after this patch is merged.

antoniofrighetto commented on llvm/lib/CodeGen/PHIEliminationUtils.cpp:
```diff
@@ -13,6 +13,19 @@
+
+  for (const MachineInstr &MI : *MBB)
+    if (MI.getOpcode() == TargetOpcode::INLINEASM_BR)
+      return true;
```
Might be wrong, though I believe that, if we were to go as follows, we would lose some instructions (e.g., SETCCr, COPY) between the INLINEASM_BR and the jump to the fall through block. Looking at the tests, there seem to be ones emitted before eliminating phi-nodes.
```cpp
for (const MachineInstr &MI : reverse(*MBB)) {
    if (MI.getOpcode() == TargetOpcode::INLINEASM_BR)
      return true;
    if (!MI.isTerminator())
      break;
  }
```

arsenm replied:
It is illegal for non-terminators to follow terminators. There cannot be any instructions between the INLINEASM_BR and the end oft he block

antoniofrighetto committed:
[PHIElimination] Introduce test for PR173396 (NFC)

antoniofrighetto committed:
[PHIElimination] Ensure correct copy emission for INLINEASM_BR indirect targets

While lowering phi-nodes to copies, subsequently to f7a53d82c0902147909f28a9295a9d00b4b27d38,
ensure that such copies are placed before any INLINEASM_BR
instructions, only if this actually happens to be the case.
If a block with an INLINEASM_BR instruction and any of its
indirect jump target are distinct blocks, the new copy on
the indirect target continues to be normally emitted prior
to the terminator.

Fixes: https://github.com/llvm/llvm-project/issues/172824.
